lib/datadog/core/diagnostics/environment_logger.rb

```diff
@@ -0,0 +1,282 @@
+# typed: true
+require 'date'
+require 'json'
+require 'rbconfig'
+
+module Datadog
+  module Core
+    module Diagnostics
+      # A holistic collection of the environment in which ddtrace is running.
+      # This logger should allow for easy reporting by users to Datadog support.
+      #
+      module EnvironmentLogger
+        class << self
+          # Outputs environment information to {Datadog.logger}.
+          # Executes only for the lifetime of the program.
+          def log!(transport_responses)
+            return if (defined?(@executed) && @executed) || !log?
+
+            @executed = true
+
+            data = EnvironmentCollector.new.collect!(transport_responses)
+            data.reject! { |_, v| v.nil? } # Remove empty values from hash output
+
+            log_environment!(data.to_json)
+            log_error!('Agent Error'.freeze, data[:agent_error]) if data[:agent_error]
+          rescue => e
+            Datadog.logger.warn("Failed to collect environment information: #{e} Location: #{Array(e.backtrace).first}")
+          end
+
+          private
+
+          def log_environment!(line)
+            Datadog.logger.info("DATADOG TRACER CONFIGURATION - #{line}")
+          end
+
+          def log_error!(type, error)
+            Datadog.logger.warn("DATADOG TRACER DIAGNOSTIC - #{type}: #{error}")
+          end
```

Comment:
SINCE THIS IS TO GO INTO CORE, SHOULD WE STILL SAY TRACER HERE?

(I'M JUST USING CAPS AS A JOKE BTW, HAPPILY SIPPING ON MY TEA WHILE REVIEWING THE PR)

Reply:
That tea must be good :smile: 
Probably should reflect just Datadog instead for now.